runtime/main/src/main/java/org/qbicc/runtime/stackwalk/JavaStackWalker.java

```diff
@@ -1,36 +1,59 @@
 package org.qbicc.runtime.stackwalk;
 
+import org.qbicc.runtime.Hidden;
+
 public class JavaStackWalker implements StackFrameVisitor {
+    private Throwable exceptionObject;
     private JavaStackFrameVisitor visitor;
     private int javaFrameCount;
 
-    private JavaStackWalker(JavaStackFrameVisitor visitor) {
+    private JavaStackWalker(Throwable exceptionObject, JavaStackFrameVisitor visitor) {
+        this.exceptionObject = exceptionObject;
         this.visitor = visitor;
-        javaFrameCount = 0;
+        this.javaFrameCount = 0;
     }
 
-    public static int getFrameCount() {
-        JavaStackWalker javaStackWalker = new JavaStackWalker(new NopVisitor());
+    @Hidden
+    public static int getFrameCount(Throwable exceptionObject) {
+        JavaStackWalker javaStackWalker = new JavaStackWalker(exceptionObject, new NopVisitor());
         StackWalker.walkStack(javaStackWalker);
         return javaStackWalker.javaFrameCount;
     }
 
-    public static void walkStack(JavaStackFrameVisitor visitor) {
-        StackFrameVisitor javaStackWalker = new JavaStackWalker(visitor);
+    @Hidden
+    public static void walkStack(Throwable exceptionObject, JavaStackFrameVisitor visitor) {
+        StackFrameVisitor javaStackWalker = new JavaStackWalker(exceptionObject, visitor);
         StackWalker.walkStack(javaStackWalker);
     }
 
+    private boolean skipFrame(int scIndex) {
+        int minfoIndex = MethodData.getMethodInfoIndex(scIndex);
+        String methodName = MethodData.getMethodName(minfoIndex);
+        String className = MethodData.getClassName(minfoIndex);
+        if (exceptionObject.getClass().getName().equals(className)) {
+            // skip frames for excepption constructor or fillInStackTrace methods
```

Comment:
The exception constructor may be on the stack more than one time; I think what we want to do here is to skip the topmost frame exception frame(s) instead.

Reply:
Fixed it.